group_utils/javascript/automation/lib/javascript/community/groupUtils.js

```diff
@@ -0,0 +1,138 @@
+/**
+ * Utilities for working with openHAB groups in JavaScript.
+ * This code is not compatible with the GraalVM JavaScript add-on.
+ *
+ * Tested to work with openHAB 3.1.0 stable.
+ * Copyright (c) 2021 Florian Hotze under MIT License
+ * @florian-h05(https://github.com/florian-h05)
+ */
+
+/**
+ * Constructor, initializes the logger.
+ */
+var GroupUtils = function () {
+  this.log = Java.type('org.slf4j.LoggerFactory').getLogger('org.openhab.model.script.Rules.GroupUtils')
+  this.log.debug('Building groupUtils instance.')
+}
+
+/**
+ * Get the direct members of a group.
+ *
+ * @param {*} group Name of the openHAB group
+ * @param {*} characteristic Which characteristic of the members you get, valid: name (default), label, state
+ */
+GroupUtils.prototype.getMembers = function (group, characteristic) {
+  this.log.debug('Getting direct members of group ' + group)
+  var groupMembers = Array()
+  itemRegistry.getItem(group).getMembers().stream().forEach(function (item) {
+    if (characteristic === 'label') {
+      groupMembers.push(item.getLabel())
+    } else if (characteristic === 'state') {
+      groupMembers.push(item.getState())
+    } else {
+      groupMembers.push(item.getName())
+    }
+  })
+  // return the array
+  return groupMembers
+}
+
+/**
+ * Get all (also childs) members of a group, the group items are excluded from this list.
+ *
+ * @param {*} group Name of the openHAB group
+ * @param {*} characteristic Which characteristic of the members you get, valid: name (default), label, state
+ */
+GroupUtils.prototype.getAllMembers = function (group, characteristic) {
+  this.log.debug('Getting all members of group ' + group)
+  var groupAllMembers = Array()
+  itemRegistry.getItem(group).getAllMembers().stream().forEach(function (item) {
+    if (characteristic === 'label') {
+      groupAllMembers.push(item.getLabel())
+    } else if (characteristic === 'state') {
+      groupAllMembers.push(item.getState())
+    } else {
+      groupAllMembers.push(item.getName())
+    }
+  })
+  // return the array
+  return groupAllMembers
+}
+
+/**
+ * Perform arithmetic operations on multiple states.
+ *
+ * @param {*} items Array of states
+ * @param {*} func Arithmetic function to perform, valid: SUM, AVG, MIN, MAX
+ */
+GroupUtils.prototype.arithmetic = function (items, func) {
```

Comment:
As with the above, I'd rather see `sumMembers`, `avgMemebers` as separate functions rather than requiring the users to pass a magic string.

You need to add error checking here. Not all members of a Group are guaranteed to be the same type and not all Item types are suitable for doing math. What if you run into an HSBType? 

What happens if some Items have a Unit of Measurement and others do not? 

Why pass an Array of items here? If this is a Group util, the function should accept the group name and the function can handle getting the members. If it is intended to be a generic function, it should be in a separate library. 

A map reduce could be used here.

```
var stats =  context.ir.getItem(group)
                                   .members()
                                   .stream()
                                   .filter(function(i) {
                                     // todo only return true if i has a state that can be used to do math
                                   })
                                   .collect(Collectors.summarizingDouble(function(i) {
                                     return i.state.toDouble(); // todo we may need to cast to Number first
                                   });

// for sum
return stats.getSum();

// for average
return stats.getAverage();

// for min
return stats.getMin();

for max
return stats.getMax();
```

In the error checking, I think there should be logs, but make them debug logs because it is perfectly acceptable to run a sum on members of a Group where some members are NULL or UNDEF.